### PR TITLE
fix RuntimeError in export_phrases (change defaultdict to dict)

### LINKED ISSUE #3031: Runtime error in phrases.py
#### Problem description

Trying to use export_phrases function on a phrases model.    
Instead getting Runtime error

#### Steps/code/corpus to reproduce

```python
from gensim.models.phrases import Phrases, ENGLISH_CONNECTOR_WORDS

documents = ["I am interested in machine learning projects", 
             "machine learning projects can be useful sometimes",
            "I love working on machine learning projects",
            "interested does not mean the same thing as likes",
            "i am interested in blockchain"]

sentence_stream = [doc.split(" ") for doc in documents]
bigrams = Phrases(sentence_stream, min_count=2, threshold=1, connector_words=ENGLISH_CONNECTOR_WORDS)
trigrams = Phrases(bigrams[sentence_stream], min_count=2, threshold=1)
trigrams.export_phrases()
```

```
RuntimeError                              Traceback (most recent call last)
<ipython-input-190-0f7e41471301> in <module>
----> 1 trigrams.export_phrases()

~\Anaconda3\lib\site-packages\gensim\models\phrases.py in export_phrases(self)
    716         """
    717         result, source_vocab = {}, self.vocab
--> 718         for token in source_vocab:
    719             unigrams = token.split(self.delimiter)
    720             if len(unigrams) < 2:

RuntimeError: dictionary changed size during iteration
```

#### Versions

Please provide the output of:

```python
Windows-10-10.0.17763-SP0
Python 3.8.3 (default, Jul  2 2020, 17:30:36) [MSC v.1916 64 bit (AMD64)]
Bits 64
NumPy 1.19.5
SciPy 1.5.2
gensim 4.0.0beta
FAST_VERSION 0
```

### PR DESCRIPTION
**Pull request by thalishsajeed**
Fixes #3031

This PR - 

* Replaces defaultdict with dict in phrases.py to avoid the Runtime error caused when defaultdict modifies its contents during runtime each time it encounters an out of vocabulary token. 
* Adds test cases to check for above bug

### CODE REVIEW

**Comment by thalishsajeed** [comment]
@piskvorky Hi, hope I've done everything right this time :) Thank you for all the help!

**Comment by piskvorky** [comment]
Use a better PR title and description please.

File: gensim/models/phrases.py [hunk 1]
```diff
@@ -412,7 +411,7 @@ def load(cls, *args, **kwargs):
             if not isinstance(word, str):
                 logger.info("old version of %s loaded, upgrading %i words in memory", cls.__name__, len(model.vocab))
                 logger.info("re-save the loaded model to avoid this upgrade in the future")
-                vocab = defaultdict(int)
+                vocab = dict()
```
Comment (piskvorky) [inline gensim/models/phrases.py:414]: More Pythonic:

```suggestion
                vocab = {}
```

(here and elsewhere)

**Review by piskvorky — APPROVED** [review]
Changes look good, thanks! But fix the PR title + description please. We use those to generate Release Notes, and auto-close affected tickets on merge.

File: gensim/models/phrases.py (hunk 1 above)
Comment (thalishsajeed) [reply gensim/models/phrases.py:414]: @piskvorky Fixed the PR+title and Descriotion. Is this acceptable? I also pushed the code to my branch after changing to `vocab = {}` but It's got getting reflected here?

Comment (piskvorky) [reply gensim/models/phrases.py:414]: I don't understand your question. If the changed code you pushed got reflected here, that's good. Except I can't see it.

Comment (mpenkov) [reply gensim/models/phrases.py:414]: @thalishsajeed I don't see the changes here, so it's likely you either pushed to some other branch, or not at all. Please double-check.

Comment (thalishsajeed) [reply gensim/models/phrases.py:414]: @mpenkov  You can see 7 commits on my branch but only 6 commits are being reflected here. Any idea what I'd be doing wrong? 

`my develop branch`
Capture

Comment (piskvorky) [reply gensim/models/phrases.py:414]: Might be a Github (Microsoft) issue. I remember seeing a discussion around flaky PRs [here](https://news.ycombinator.com/item?id=26066365) (with some resolution tips too).

Comment (mpenkov) [reply gensim/models/phrases.py:414]: Yeah, I opened a ticket with support, but haven't heard back from them.

@thalishsajeed perhaps try force-pushing?

Comment (thalishsajeed) [reply gensim/models/phrases.py]: @mpenkov Looks like my changes are reflecting now!

**Review by piskvorky — APPROVED** [review]

**Review by mpenkov — APPROVED** [review]

**Comment by mpenkov** [comment]
Merged. Thank you @thalishsajeed and congrats on your first contribution to gensim!

**Comment by thalishsajeed** [comment]
@mpenkov  Thank you! :) Looking forward to gensim 4.0

**Pull request merged**